{python313Packages.pyliblo,raysession}: drop

ArchercatNEO opened the pull request:
Fixes #389149
Maybe fixes #320429, #323195 (dropping raysession)

Drops pyliblo which hasn't had activity in 10 years and fails to build under python 3.13
Raysession has been removed because of its required dependency on pyliblo.
Carla only has an optional dependency of pyliblo which has been disabled.
Please comment if you know this significantly impacts carla's functionality.

An alternative is to package a fork at https://github.com/gesellkammer/pyliblo3. This fork seems reputable enough as it backed by Debian, Ubuntu and is mentioned as an alternative in both of the mentioned apps.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

emilazy reviewed (commented):
cc @orivej @Toboi – apparently RaySession currently fails to launch on unstable?

Dropping `pylibio` entirely and packaging the fork both seem like sensible options, depending on whether we’re keeping RaySession and whether the functionality in Carla is valuable. Let’s wait a bit to hear back about RaySession.

You’ll need to add a `throw` to `python-aliases.nix`.

ArchercatNEO commented:
Since this is a python package the compilation suceeds just fine but once it runs there's a missing `cgitb` module

```
❯ ./result/bin/raysession
Traceback (most recent call last):
  File "/nix/store/szsxkpa9g3l970ki3cddp21j2dxl8sb5-raysession-0.14.4/share/raysession/src/gui/.raysession.py-wrapped", line 16, in <module>
    from gui_session import SignaledSession
  File "/nix/store/szsxkpa9g3l970ki3cddp21j2dxl8sb5-raysession-0.14.4/share/raysession/src/gui/gui_session.py", line 6, in <module>
    from client_properties_dialog import NsmClientPropertiesDialog
  File "/nix/store/szsxkpa9g3l970ki3cddp21j2dxl8sb5-raysession-0.14.4/share/raysession/src/gui/client_properties_dialog.py", line 12, in <module>
    from child_dialogs import ChildDialog
  File "/nix/store/szsxkpa9g3l970ki3cddp21j2dxl8sb5-raysession-0.14.4/share/raysession/src/gui/child_dialogs.py", line 23, in <module>
    import ui.save_template_session
  File "/nix/store/szsxkpa9g3l970ki3cddp21j2dxl8sb5-raysession-0.14.4/share/raysession/src/gui/ui/save_template_session.py", line 109, in <module>
    from surclassed_widgets import FakeToolButton
  File "/nix/store/szsxkpa9g3l970ki3cddp21j2dxl8sb5-raysession-0.14.4/share/raysession/src/gui/surclassed_widgets.py", line 12, in <module>
    from patchbay import filter_frame, PatchGraphicsView
  File "/nix/store/szsxkpa9g3l970ki3cddp21j2dxl8sb5-raysession-0.14.4/share/raysession/src/gui/patchbay/__init__.py", line 1, in <module>
    from .patchbay_manager import PatchbayManager, patchcanvas
  File "/nix/store/szsxkpa9g3l970ki3cddp21j2dxl8sb5-raysession-0.14.4/share/raysession/src/gui/patchbay/patchbay_manager.py", line 14, in <module>
    from .patchcanvas import patchcanvas, PortType, PortSubType, PortMode
  File "/nix/store/szsxkpa9g3l970ki3cddp21j2dxl8sb5-raysession-0.14.4/share/raysession/src/gui/patchbay/patchcanvas/__init__.py", line 32, in <module>
    from .patchcanvas import *
  File "/nix/store/szsxkpa9g3l970ki3cddp21j2dxl8sb5-raysession-0.14.4/share/raysession/src/gui/patchbay/patchcanvas/patchcanvas.py", line 51, in <module>
    from .box_widget import BoxWidget
  File "/nix/store/szsxkpa9g3l970ki3cddp21j2dxl8sb5-raysession-0.14.4/share/raysession/src/gui/patchbay/patchcanvas/box_widget.py", line 19, in <module>
    from .box_widget_moth import BoxWidgetMoth, UnwrapButton, TitleLine
  File "/nix/store/szsxkpa9g3l970ki3cddp21j2dxl8sb5-raysession-0.14.4/share/raysession/src/gui/patchbay/patchcanvas/box_widget_moth.py", line 50, in <module>
    from .portgroup_widget import PortgroupWidget
  File "/nix/store/szsxkpa9g3l970ki3cddp21j2dxl8sb5-raysession-0.14.4/share/raysession/src/gui/patchbay/patchcanvas/portgroup_widget.py", line 22, in <module>
    from cgitb import text
ModuleNotFoundError: No module named 'cgitb'

```

dotlambda reviewed (commented):
You can pin raysession to `python312Packages` instead.
Please add a comment with the reason though: https://github.com/NixOS/nixpkgs/pull/416728.
Also, one commit per package.

ArchercatNEO committed:
raysession: drop

Maybe fixes #320429, #323195

Raysession has a required dependency on the unmaintained pyliblo.
Pyliblo is being dropped which requires that raysession be dropped too.

If there is still demand for raysession an alternative is to package
https://github.com/gesellkammer/pyliblo3 which raysession itself
endorses and pinning it to python 3.12.
Pinning is necessary because the current latest release (14.4) depends
on `cgitb` which was removed in python 3.13.
Upstream is aware of the issue and has fixed in master but no release
contains the fix as of now.

ArchercatNEO committed:
python313Packages.pyliblo: drop

Fixes #389149

Drops pyliblo which hasn't had activity in 10 years and fails to build under python 3.13
Raysession has been removed because of its required dependency on pyliblo.
Carla only has an optional dependency of pyliblo which has been disabled.
Please comment if you know this significantly impacts carla's functionality.

An alternative is to package a fork at https://github.com/gesellkammer/pyliblo3.
This fork seems reputable enough as it backed by Debian, Ubuntu and is mentioned as an alternative in both of the mentioned apps.

ArchercatNEO commented:
> You can pin raysession to python312Packages instead.

A clarification: `cgitb` was a standard python module but it got deprecated and removed in python 3.13 (hence the failure).

Upstream is aware of this and has been fixed in [master](https://github.com/Houston4444/HoustonPatchbay/blob/main/source/patchbay/patchcanvas/portgroup_widget.py) 
but the [latest release (14.4)](https://github.com/Houston4444/HoustonPatchbay/blob/0716077f4713b0c627d42a537ed729d9e282a181/patchbay/patchcanvas/portgroup_widget.py) doesn't contain the fix so to fix raysession we could package the fork *and* pin raysession to python312 until a new release is made. That is if there is still demand for raysession

dotlambda commented:
> > You can pin raysession to python312Packages instead.
> 
> A clarification: `cgitb` was a standard python module but it got deprecated and removed in python 3.13 (hence the failure).

So another thing that would be solved by using `python312Packages`.
There's no reason to drop raysession.

dotlambda commented:
Use `python312Packages` for raysession and everything is fine. No need to mess with pyliblo.

ArchercatNEO commented:
I am sorry for having brought up cgitb because it seems that that made it look like I’m interested in solving a problem with raysession. I am making this PR because for some reason pyliblo’s build failure is causing system rebuilds to fail even for people who don’t depend on it.

Would pinning raysession to 3.12 cause python313Packages.pyliblo to either not exist or succeed compilation? I don’t understand a method by which it could which is why I haven’t tried pinning raysession yet.

Once I get home I will try to pin raysession to see if it fixes pyliblo. If I am late and pyliblo already compiles in the master branch then this PR doesn’t have value and could just be closed

dotlambda commented:
> Would pinning raysession to 3.12 cause python313Packages.pyliblo to either not exist or succeed compilation?

No of course not.



> I am making this PR because for some reason pyliblo’s build failure is causing system rebuilds to fail even for people who don’t depend on it.

Those people must be using something that depends on pybliblo, which can only be raysession or carla. So fixing those two is sufficient.

dotlambda commented:
https://github.com/Houston4444/RaySession/blob/4615d1a9de5573acaca10f31fed54495499a9c7a/INSTALL.md?plain=1#L72

dotlambda commented on pkgs/applications/audio/carla/default.nix:
```diff
@@ -47,7 +47,6 @@ stdenv.mkDerivation (finalAttrs: {
     with python3Packages;
     [
       rdflib
-      pyliblo
```
Doesn't this break carla?
See also https://github.com/falkTX/Carla/issues/1052

ArchercatNEO replied:
Surpringly no. I managed to compile, run and get audio out of Carla without pyliblo. Unfortunately I do not have enough experience with Carla or OSC and could not say if any important functionality is lost from this

ArchercatNEO commented:
> https://github.com/Houston4444/RaySession/blob/4615d1a9de5573acaca10f31fed54495499a9c7a/INSTALL.md?plain=1#L72

What do you intend by this? The fork mentioned is the one I mentioned in option 2. We do not currently package the fork and option 2 is precisely to replace pyliblo with its fork. It would have to be "messing with the pyliblo expression" since the src needs to change. I have tested it and fixes the build problem on python 3.13 but obviously won't fix the missing `cgitb` module (which would be its own PR).

If you support moving to the fork I can make a different PR to do just that and we can close this one without merge.

ArchercatNEO commented:
I have created 2 branches: one is for porting to the pyliblo3 fork and one for pinning raysession to python312Packages.

https://github.com/ArchercatNEO/nixpkgs/tree/pylib-port
https://github.com/ArchercatNEO/nixpkgs/tree/raysession-pin

The one with only the raysession pin compiles successfully and opens but I don't have JACK running so I didn't test beyond that. However, I don't have any of pyliblo, carla or raysession installed and the pyliblo build failure still caused my rebuild to fail somehow which leads me to believe that even if no app in nixpkgs explicitly uses `python313Packages.liblo` we should still ensure it actually builds which is what the other branch is for.

Please tell me if any of these work for you because my primary goal is still to solve the `python313Packages.liblo` build failure even if it seems to not matter to you. Even if none of these branches seem good could you at least explain why we aren't trying to fix the `python313Packages.liblo` build fauilure? 

If they seem alright I can set up PRs however is needed.

dotlambda commented:
Packaging the fork sounds like a good idea.

ArchercatNEO commented:
Superseded by #418925
